openpifpaf/plugins/apollocar3d/apollocar3d/apollo_to_coco.py

```diff
@@ -0,0 +1,291 @@
+        """Process single instance"""
+        all_kps = np.array(data)  # [#, x, y]
+
+        # Enlarge box
+        box_tight = [np.min(all_kps[:, 1]), np.min(all_kps[:, 2]), np.max(all_kps[:, 1]), np.max(all_kps[:, 2])]
+        w, h = box_tight[2] - box_tight[0], box_tight[3] - box_tight[1]
+        x_o = max(box_tight[0] - (w / 10), 0)
+        y_o = max(box_tight[1] - (h / 10), 0)
+        x_i = min(x_o + (w / 4) + w, im_size[0])
+        y_i = min(y_o + (h / 4) + h, im_size[1])
+        box = [int(x_o), int(y_o), int(x_i - x_o), int(y_i - y_o)]  # (x, y, w, h)
+
+        kps, num = self._transform_keypoints(all_kps)
+        txt_id = os.path.splitext(txt_path.split(sep='_')[-1])[0]
+        car_id = int(str(im_id) + str(int(txt_id)))  # include at the end of the number the specific annotation id
+        self.json_file["annotations"].append({
+            'image_id': im_id,
+            'category_id': 1,
+            'iscrowd': 0,
+            'id': car_id,
+            'area': box[2] * box[3],
+            'bbox': box,
+            'num_keypoints': num,
+            'keypoints': kps,
+            'segmentation': []})
+        # Stats
+        for num in data[0]:
+            cnt_kps[num] += 1
+        return cnt_kps
+
+    def _transform_keypoints(self, kps):
+        """
+        Map, filter keypoints and add visibility
+        :array of [[#, x, y], ...]
+        :return  [x, y, visibility, x, y, visibility, .. ]
+        """
+        kps_out = np.zeros((self.num_kps, 3))
+        cnt = 0
+        for kp in kps:
+            n = self.map_sk[int(kp[0])]
+            if n < 100:  # Filter extra keypoints
+                kps_out[n, 0] = kp[1]
+                kps_out[n, 1] = kp[2]
+                kps_out[n, 2] = 2
+                cnt += 1
+        kps_out = list(kps_out.reshape((-1,)))
+        return kps_out, cnt
+
+    def initiate_json(self):
+        """
+        Initiate Json for training and val phase
+        """
+        self.json_file["info"] = dict(url="https://github.com/vita-epfl/openpifpaf",
+                                      date_created=time.strftime("%a, %d %b %Y %H:%M:%S +0000", time.localtime()),
+                                      description="Conversion of ApolloCar3D dataset into MS-COCO format")
+
+        self.json_file["categories"] = [dict(name='car',
+                                             id=1,
+                                             skeleton=CAR_SKELETON,
+                                             supercategory='car',
+                                             keypoints=CAR_KEYPOINTS)]
+        self.json_file["images"] = []
+        self.json_file["annotations"] = []
+
+
+def histogram(cnt_kps):
+    if plt is None:
+        raise Exception('please install matplotlib')
+    bins = np.arange(len(cnt_kps))
+    data = np.array(cnt_kps)
+    plt.figure(1)
+    plt.bar(bins, data)
+    plt.xticks(np.arange(len(cnt_kps), step=5))
+    plt.show()
+
+
+def make_new_directory(dir_out):
```

Comment:
I would never remove the user's directories. Feel free to add timestamps to file names to make them unique or something similar but don't delete anything on the hard drive. Even if that means things get messy.

Reply:
You are right. I added an assert for non empty directory, as this is not expected